solr/solrj/src/java/org/apache/solr/common/cloud/LazyNodePropsProvider.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.common.cloud;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.solr.client.solrj.impl.CloudSolrClient;
+import org.apache.solr.common.NavigableObject;
+import org.apache.solr.common.util.Utils;
+
+/**
+ * Fetch lazily and cache a node's system properties
+ *
+ */
+public class LazyNodePropsProvider implements NodePropsProvider {
+  private volatile boolean isClosed = false;
+  private Map<String ,Map<String, Object>> nodeVsTagsCache = new ConcurrentHashMap<>();
+  private final String thisNode;
+  private final ZkStateReader zkStateReader;
+  private final CloudSolrClient cloudSolrClient;
+
+  public LazyNodePropsProvider(String thisNode, CloudSolrClient cloudSolrClient, ZkStateReader zkStateReader) {
+    this.zkStateReader = zkStateReader;
+    this.cloudSolrClient = cloudSolrClient;
+    this.thisNode = thisNode;
+    zkStateReader.registerLiveNodesListener((oldNodes, newNodes) -> {
+      for (String n : oldNodes) {
+        if(Objects.equals(n, thisNode)) continue;
```

Comment:
Can you explain the special case on why we always want to keep `thisNode` in the nodeVsTagsCache? :)